pwndbg/color/disasm.py

```diff
@@ -22,67 +23,123 @@
+#  Example: `✔ je     _IO_file_xsputn+341`. Inline symbol replacements made. No annotation or branch targets shown.
+def one_instruction(ins: PwndbgInstruction) -> str:
+    asm = ins.asm_string
 
     # Highlight the current line if enabled
     if pwndbg.gdblib.config.highlight_pc and ins.address == pwndbg.gdblib.regs.pc:
         asm = C.highlight(asm)
 
-    # tl;dr is a branch?
-    if ins.target not in (None, ins.address + ins.size):
-        sym = pwndbg.gdblib.symbol.get(ins.target) or None
-        target = M.get(ins.target)
-        const = ins.target_const
-        hextarget = hex(ins.target)
-        hexlen = len(hextarget)
-
-        # If it's a constant expression, color it directly in the asm.
-        if const:
-            asm = f"{ljust_colored(asm, 36)} <{target}>"
-            asm = asm.replace(hex(ins.target), sym or target)
-
-        # It's not a constant expression, but we've calculated the target
-        # address by emulation or other means (for example showing ret instruction target)
-        elif sym:
-            asm = f"{ljust_colored(asm, 36)} <{target}; {sym}>"
-
-        # We were able to calculate the target, but there is no symbol
-        # name for it.
-        else:
-            asm += f"<{(target)}>"
-
-    # not a branch
-    elif ins.symbol:
-        if is_branch and not ins.target:
-            asm = f"{asm} <{ins.symbol}>"
-
-            # XXX: not sure when this ever happens
-            asm += "<-- file a pwndbg bug for this"
-        else:
-            inlined_sym = asm.replace(hex(ins.symbol_addr), ins.symbol)
+    is_call_or_jump = ins.groups_set & capstone_branch_groups
 
-            # display symbol as mem text if no inline replacement was made
-            mem_text = ins.symbol if inlined_sym == asm else None
-            asm = f"{ljust_colored(inlined_sym, 36)} <{M.get(ins.symbol_addr, mem_text)}>"
-
-    # Style the instruction mnemonic if it's a branch instruction.
-    if is_branch:
+    # Style the instruction mnemonic if it's a call/jump instruction.
+    if is_call_or_jump:
         asm = asm.replace(ins.mnemonic, c.branch(ins.mnemonic), 1)
 
     # If we know the conditional is taken, mark it as taken.
-    if ins.condition is None:
-        asm = "  " + asm
-    elif ins.condition:
+    if ins.condition == InstructionCondition.TRUE or ins.is_conditional_jump_taken:
         asm = on("✔ ") + asm
     else:
         asm = "  " + asm
 
     return asm
+
+
+# To making the padding visually nicer, the following padding scheme is used for annotations:
+# All instructions in a group will have the same amount of left-adjusting spaces, so they are aligned.
+# A group is defined as a sequence of instructions surrounded by instructions that can change the instruction pointer.
+def instructions_and_padding(instructions: list[PwndbgInstruction]) -> list[str]:
+    assembly = [one_instruction(i) for i in instructions]
```

Comment:
Since this is enumerated, we don't need to render the list here, we can execute it on demand.
```suggestion
    assembly = (one_instruction(i) for i in instructions)
```

In practice, this should also be called in the loop below